michbl/assignment2/BinaryTree.hpp

```diff
@@ -0,0 +1,194 @@
+#ifndef BINARYTREE_HPP
+#define BINARYTREE_HPP
+
+#include <iostream>
+using namespace std;
+
+// Assuming T is comparable and printable
+// Assuming keys are unique
+
+template <typename T>
+struct Node {
+    T key;
+    Node<T> *left = nullptr;
+    Node<T> *right = nullptr;
+};
+
+template <typename T>
+class BinaryTree {
+private:
+    Node<T> *root;
+
+    void copyTree(Node<T> *parent, Node<T> *newNode, string side) {
+        if (newNode == nullptr) {
+            return;
+        }
+
+        Node<T> *child = new Node<T>;
+        child->key = newNode->key;
+        if (side == "left") {
+            parent->left = child;
+        }
+        else {
+            parent->right = child;
+        }
+        copyTree(child, newNode->left, "left");
+        copyTree(child, newNode->right, "right");
+    }
+
+    void insert(Node<T> *node, T key, T parentKey, bool &success) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == parentKey) {
+            if (node->left == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->left = child;
+                success = true;
+                return;
+            }
+            else if (node->right == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->right = child;
+                success = true;
+                return;
+            }
+            else { // parent already has 2 children
+                return;
+            }
+        }
+        insert(node->left, key, parentKey, success);
+        if (success) {
+            return;
+        }
+        insert(node->right, key, parentKey, success);
+    }
+
+    void destroyTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        destroyTree(node->left);
+        destroyTree(node->right);
+        delete node;
+    }
+
+    void printTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        printTree(node->left);
+        printTree(node->right);
+        cout << node->key << " ";
+    }
+
+    void printAncestors(Node<T> *node, T key, bool &found) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == key) {
+            found = true;
+            return;
+        }
+        if (!found) {

```

Comment:
At this point `found` is always false, so this `if` is redundant.

-----
Additionally, you can use the return value instead of `found` output parameter. It can help the function be more concise:
```C++
bool found = printAncestors(node->left, key);
if (!found) {
  found = printAncestors(node->right, key);t
}
if (found) {
    cout << node->key << ", " << endl;  // I got lazy about printing the trailing comma :]
    return true;
}
return false;
```
Or even more concise:
```C++
if (printAncestors(node->left, key) || printAncestors(node->right, key)) {
    // The key is found in our left or right subtree.
    cout << node->key << ", " << endl;
    return true;
}
return false;
```
The latter is shorter and has the same logic; but has the downside that `printAncestors` have side effects (prints the value). Its function name contains `print` so it helps user not to overlook it.